File: go/examples/calculator/calculator.go
package main

import (
	"math"
	"os"
	"strconv"
	"strings"
	"unicode"

	uimd "uimd"
)

const (
	calculatorDisplayZero  = "0"
	calculatorDisplayError = "Error"
	calculatorPercent      = 100.0
)

type expressionParser struct {
	input    string
	position int
}

func newExpressionParser(input string) *expressionParser {
	return &expressionParser{input: input}
}

func (parser *expressionParser) parse() (float64, bool) {
	value, ok := parser.expression()
	if !ok {
		return 0, false
	}
	parser.skipSpaces()
	return value, parser.position == len(parser.input)
}

func (parser *expressionParser) expression() (float64, bool) {
	value, ok := parser.term()
	if !ok {
		return 0, false
	}
	for {
		parser.skipSpaces()
		if parser.match('+') {
			next, ok := parser.term()
			if !ok {
				return 0, false
			}
			value += next
		} else if parser.match('-') {
			next, ok := parser.term()
			if !ok {
				return 0, false
			}
			value -= next
		} else {
			return value, true
		}
	}
}

func (parser *expressionParser) term() (float64, bool) {
	value, ok := parser.factor()
	if !ok {
		return 0, false
	}
	for {
		parser.skipSpaces()
		if parser.match('*') {
			next, ok := parser.factor()
			if !ok {
				return 0, false
			}
			value *= next
		} else if parser.match('/') {
			next, ok := parser.factor()
			if !ok {
				return 0, false
			}
			value /= next
		} else {
			return value, true
		}
	}
}

func (parser *expressionParser) factor() (float64, bool) {
	parser.skipSpaces()
	if parser.match('+') {
		return parser.factor()
	}
	if parser.match('-') {
		value, ok := parser.factor()
		return -value, ok
	}
	return parser.number()
}

func (parser *expressionParser) number() (float64, bool) {
	parser.skipSpaces()
	start := parser.position
	for parser.position < len(parser.input) {
		ch := rune(parser.input[parser.position])
		if !unicode.IsDigit(ch) && ch != '.' {
			break
		}
		parser.position++
	}
	if start == parser.position {
		return 0, false
	}
	value, err := strconv.ParseFloat(parser.input[start:parser.position], 64)
	return value, err == nil
}

func (parser *expressionParser) match(value byte) bool {
	if parser.position < len(parser.input) && parser.input[parser.position] == value {
		parser.position++
		return true
	}
	return false
}

func (parser *expressionParser) skipSpaces() {
	for parser.position < len(parser.input) && unicode.IsSpace(rune(parser.input[parser.position])) {
		parser.position++
	}
}

type Calculator struct {
	*CalculatorUI
	expression  string
	afterResult bool
}

func NewCalculator() *Calculator {
	app := &Calculator{
		CalculatorUI: NewCalculatorUI(),
		expression:   calculatorDisplayZero,
	}
	app.SetEventHandler(app)
	app.refresh()
	return app
}

func (app *Calculator) RuntimeOptions() uimd.GeneratedWindowRuntimeOptions {
	options := app.CalculatorUI.RuntimeOptions()
	options.OnKey = app.key
	return options
}

func (app *Calculator) OnN0Click() { app.button("n0") }
func (app *Calculator) OnN1Click() { app.button("n1") }
func (app *Calculator) OnN2Click() { app.button("n2") }
func (app *Calculator) OnN3Click() { app.button("n3") }
func (app *Calculator) OnN4Click() { app.button("n4") }
func (app *Calculator) OnN5Click() { app.button("n5") }
func (app *Calculator) OnN6Click() { app.button("n6") }
func (app *Calculator) OnN7Click() { app.button("n7") }
func (app *Calculator) OnN8Click() { app.button("n8") }
func (app *Calculator) OnN9Click() { app.button("n9") }
func (app *Calculator) OnDtClick() { app.button("dt") }
func (app *Calculator) OnAdClick() { app.button("ad") }
func (app *Calculator) OnMnClick() { app.button("mn") }
func (app *Calculator) OnMlClick() { app.button("ml") }
func (app *Calculator) OnDvClick() { app.button("dv") }
func (app *Calculator) OnSbClick() { app.button("sb") }
func (app *Calculator) OnAcClick() { app.button("ac") }
func (app *Calculator) OnDlClick() { app.button("dl") }
func (app *Calculator) OnPrClick() { app.button("pr") }

func (app *Calculator) button(name string) {
	switch name {
	case "n0":
		app.digit('0')
	case "n1":
		app.digit('1')
	case "n2":
		app.digit('2')
	case "n3":
		app.digit('3')
	case "n4":
		app.digit('4')
	case "n5":
		app.digit('5')
	case "n6":
		app.digit('6')
	case "n7":
		app.digit('7')
	case "n8":
		app.digit('8')
	case "n9":
		app.digit('9')
	case "dt":
		app.digit('.')
	case "ad":
		app.operation('+')
	case "mn":
		app.operation('-')
	case "ml":
		app.operation('*')
	case "dv":
		app.operation('/')
	case "sb":
		app.equalsPressed()
	case "ac":
		app.reset()
	case "dl":
		app.delete()
	case "pr":
		app.percent()
	}
}

func (app *Calculator) key(key string) bool {
	if len(key) == 1 && key[0] >= '0' && key[0] <= '9' {
		app.digit(rune(key[0]))
		return true
	}
	switch key {
	case ".":
		app.digit('.')
	case "+", "-", "*", "/":
		app.operation(rune(key[0]))
	case "=":
		app.equalsPressed()
	case "%":
		app.percent()
	case "Backspace":
		app.delete()
	default:
		return false
	}
	return true
}

func (app *Calculator) digit(digit rune) {
	if app.afterResult || app.expression == calculatorDisplayError {
		app.expression = ""
		app.afterResult = false
	}
	if digit == '.' {
		app.appendDecimalPoint()
	} else if app.expression == calculatorDisplayZero {
		app.expression = string(digit)
	} else if current := app.currentNumberText(); current == "0" || current == "-0" {
		app.replaceCurrentNumber(strings.Replace(current, calculatorDisplayZero, string(digit), 1))
	} else {
		app.expression += string(digit)
	}
	app.refresh()
}

func (app *Calculator) operation(operation rune) {
	if app.expression == calculatorDisplayError {
		app.expression = calculatorDisplayZero
	}
	if app.expression == "" {
		if operation == '-' {
			app.expression = string(operation)
			app.afterResult = false
			app.refresh()
		}
		return
	}
	if isOperator(rune(app.expression[len(app.expression)-1])) {
		if operation == '-' && app.expression[len(app.expression)-1] != '-' {
			app.expression += string(operation)
		} else {
			app.expression = strings.TrimRight(app.expression, "+-*/") + string(operation)
		}
	} else {
		app.expression += string(operation)
	}
	app.afterResult = false
	app.refresh()
}

func (app *Calculator) equalsPressed() {
	app.compute()
}

func (app *Calculator) delete() {
	if app.afterResult || app.expression == calculatorDisplayError {
		app.reset()
		return
	}
	if app.expression != "" {
		app.expression = app.expression[:len(app.expression)-1]
	}
	if app.expression == "" {
		app.expression = calculatorDisplayZero
	}
	app.refresh()
}

func (app *Calculator) percent() {
	current := app.currentNumberText()
	if current == "" {
		return
	}
	value, err := strconv.ParseFloat(current, 64)
	if err != nil {
		app.showError()
		return
	}
	app.replaceCurrentNumber(formatCalculator(value / calculatorPercent))
	app.afterResult = false
	app.refresh()
}

func (app *Calculator) compute() {
	result, ok := newExpressionParser(app.expression).parse()
	if !ok || math.IsInf(result, 0) || math.IsNaN(result) {
		app.showError()
		return
	}
	app.expression = formatCalculator(result)
	app.afterResult = true
	app.refresh()
}

func (app *Calculator) reset() {
	app.expression = calculatorDisplayZero
	app.afterResult = false
	app.refresh()
}

func (app *Calculator) refresh() {
	if app.expression == "" {
		app.Display.SetText(calculatorDisplayZero)
		return
	}
	app.Display.SetText(app.expression)
}

func (app *Calculator) showError() {
	app.expression = calculatorDisplayError
	app.afterResult = true
	app.refresh()
}

func (app *Calculator) appendDecimalPoint() {
	current := app.currentNumberText()
	if strings.Contains(current, ".") {
		return
	}
	if app.expression == "" || isOperator(rune(app.expression[len(app.expression)-1])) {
		app.expression += calculatorDisplayZero + "."
	} else {
		app.expression += "."
	}
}

func (app *Calculator) currentNumberSpan() (int, int) {
	end := len(app.expression)
	if end == 0 || isOperator(rune(app.expression[end-1])) {
		return end, end
	}
	start := end
	for start > 0 {
		ch := rune(app.expression[start-1])
		if !unicode.IsDigit(ch) && ch != '.' {
			break
		}
		start--
	}
	if start > 0 && app.expression[start-1] == '-' && (start == 1 || isOperator(rune(app.expression[start-2]))) {
		start--
	}
	return start, end
}

func (app *Calculator) currentNumberText() string {
	start, end := app.currentNumberSpan()
	return app.expression[start:end]
}

func (app *Calculator) replaceCurrentNumber(value string) {
	start, end := app.currentNumberSpan()
	app.expression = app.expression[:start] + value + app.expression[end:]
}

func isOperator(value rune) bool {
	return value == '+' || value == '-' || value == '*' || value == '/'
}

func formatCalculator(value float64) string {
	rounded := math.Round(value)
	if math.Abs(value-rounded) < math.SmallestNonzeroFloat64*math.Max(1.0, math.Abs(value)) {
		return strconv.FormatFloat(rounded, 'f', 0, 64)
	}
	return strconv.FormatFloat(value, 'g', 10, 64)
}

func main() {
	code := uimd.RunGeneratedAppMain(func() int {
		app := NewCalculator()
		return uimd.RunGeneratedWindow(app, app.RuntimeOptions(), os.Args)
	})
	os.Exit(code)
}

File: rust/examples/calculator/calculator.rs
mod calculator_ui;

use calculator_ui::*;

const DISPLAY_ZERO: &str = "0";
const DISPLAY_ERROR: &str = "Error";
const PERCENT_FACTOR: f64 = 100.0;

struct ExpressionParser<'a>
{
    input: &'a [u8],
    position: usize,
}

impl<'a> ExpressionParser<'a>
{
    fn new(input: &'a str) -> Self
    {
        Self { input: input.as_bytes(), position: 0 }
    }

    fn parse(mut self) -> Option<f64>
    {
        let value = self.expression()?;
        self.skip_spaces();
        (self.position == self.input.len()).then_some(value)
    }

    fn expression(&mut self) -> Option<f64>
    {
        let mut value = self.term()?;
        loop
        {
            self.skip_spaces();
            if self.matches(b'+')
            {
                value += self.term()?;
            }
            else if self.matches(b'-')
            {
                value -= self.term()?;
            }
            else
            {
                return Some(value);
            }
        }
    }

    fn term(&mut self) -> Option<f64>
    {
        let mut value = self.factor()?;
        loop
        {
            self.skip_spaces();
            if self.matches(b'*')
            {
                value *= self.factor()?;
            }
            else if self.matches(b'/')
            {
                value /= self.factor()?;
            }
            else
            {
                return Some(value);
            }
        }
    }

    fn factor(&mut self) -> Option<f64>
    {
        self.skip_spaces();
        if self.matches(b'+')
        {
            return self.factor();
        }
        if self.matches(b'-')
        {
            return self.factor().map(|value| -value);
        }
        self.number()
    }

    fn number(&mut self) -> Option<f64>
    {
        self.skip_spaces();
        let start = self.position;
        while self.position < self.input.len()
            && (self.input[self.position].is_ascii_digit() || self.input[self.position] == b'.')
        {
            self.position += 1;
        }
        if start == self.position
        {
            return None;
        }
        std::str::from_utf8(&self.input[start..self.position]).ok()?.parse().ok()
    }

    fn matches(&mut self, value: u8) -> bool
    {
        if self.input.get(self.position) == Some(&value)
        {
            self.position += 1;
            true
        }
        else
        {
            false
        }
    }

    fn skip_spaces(&mut self)
    {
        while self.input.get(self.position).is_some_and(u8::is_ascii_whitespace)
        {
            self.position += 1;
        }
    }
}

struct CalculatorApp
{
    expression: String,
    after_result: bool,
}

impl CalculatorApp
{
    fn new(ui: &mut CalculatorUI) -> Self
    {
        let app = Self { expression: DISPLAY_ZERO.to_string(), after_result: false };
        app.refresh(ui);
        app
    }

    fn digit(&mut self, ui: &mut CalculatorUI, digit: char)
    {
        if self.after_result || self.expression == DISPLAY_ERROR
        {
            self.expression.clear();
            self.after_result = false;
        }
        if digit == '.'
        {
            self.append_decimal_point();
        }
        else if self.expression == DISPLAY_ZERO
        {
            self.expression = digit.to_string();
        }
        else
        {
            let current = self.current_number_text();
            if current == "0" || current == "-0"
            {
                self.replace_current_number(&current.replacen(DISPLAY_ZERO, &digit.to_string(), 1));
            }
            else
            {
                self.expression.push(digit);
            }
        }
        self.refresh(ui);
    }

    fn operation(&mut self, ui: &mut CalculatorUI, operation: char)
    {
        if self.expression == DISPLAY_ERROR
        {
            self.expression = DISPLAY_ZERO.to_string();
        }
        if self.expression.is_empty()
        {
            if operation == '-'
            {
                self.expression.push(operation);
                self.after_result = false;
                self.refresh(ui);
            }
            return;
        }
        if self.expression.chars().last().is_some_and(is_operator)
        {
            if operation == '-' && !self.expression.ends_with('-')
            {
                self.expression.push(operation);
            }
            else
            {
                while self.expression.chars().last().is_some_and(is_operator)
                {
                    self.expression.pop();
                }
                self.expression.push(operation);
            }
        }
        else
        {
            self.expression.push(operation);
        }
        self.after_result = false;
        self.refresh(ui);
    }

    fn equals(&mut self, ui: &mut CalculatorUI)
    {
        match ExpressionParser::new(&self.expression).parse()
        {
            Some(result) if result.is_finite() =>
            {
                self.expression = format_calculator(result);
                self.after_result = true;
                self.refresh(ui);
            }
            _ => self.show_error(ui),
        }
    }

    fn reset(&mut self, ui: &mut CalculatorUI)
    {
        self.expression = DISPLAY_ZERO.to_string();
        self.after_result = false;
        self.refresh(ui);
    }

    fn delete(&mut self, ui: &mut CalculatorUI)
    {
        if self.after_result || self.expression == DISPLAY_ERROR
        {
            self.reset(ui);
            return;
        }
        self.expression.pop();
        if self.expression.is_empty()
        {
            self.expression = DISPLAY_ZERO.to_string();
        }
        self.refresh(ui);
    }

    fn percent(&mut self, ui: &mut CalculatorUI)
    {
        let current = self.current_number_text();
        let Ok(value) = current.parse::<f64>() else
        {
            self.show_error(ui);
            return;
        };
        self.replace_current_number(&format_calculator(value / PERCENT_FACTOR));
        self.after_result = false;
        self.refresh(ui);
    }

    fn refresh(&self, ui: &mut CalculatorUI)
    {
        ui.display.borrow_mut().set_text(if self.expression.is_empty()
        {
            DISPLAY_ZERO
        }
        else
        {
            &self.expression
        });
    }

    fn show_error(&mut self, ui: &mut CalculatorUI)
    {
        self.expression = DISPLAY_ERROR.to_string();
        self.after_result = true;
        self.refresh(ui);
    }

    fn append_decimal_point(&mut self)
    {
        if self.current_number_text().contains('.')
        {
            return;
        }
        if self.expression.is_empty() || self.expression.chars().last().is_some_and(is_operator)
        {
            self.expression.push_str("0.");
        }
        else
        {
            self.expression.push('.');
        }
    }

    fn current_number_span(&self) -> (usize, usize)
    {
        let bytes = self.expression.as_bytes();
        let end = bytes.len();
        if end == 0 || is_operator(bytes[end - 1] as char)
        {
            return (end, end);
        }
        let mut start = end;
        while start > 0 && (bytes[start - 1].is_ascii_digit() || bytes[start - 1] == b'.')
        {
            start -= 1;
        }
        if start > 0
            && bytes[start - 1] == b'-'
            && (start == 1 || is_operator(bytes[start - 2] as char))
        {
            start -= 1;
        }
        (start, end)
    }

    fn current_number_text(&self) -> String
    {
        let (start, end) = self.current_number_span();
        self.expression[start..end].to_string()
    }

    fn replace_current_number(&mut self, value: &str)
    {
        let (start, end) = self.current_number_span();
        self.expression.replace_range(start..end, value);
    }
}

impl CalculatorUIEvents for CalculatorApp
{
    fn on_n0_click(&mut self, ui: &mut CalculatorUI) { self.digit(ui, '0'); }
    fn on_n1_click(&mut self, ui: &mut CalculatorUI) { self.digit(ui, '1'); }
    fn on_n2_click(&mut self, ui: &mut CalculatorUI) { self.digit(ui, '2'); }
    fn on_n3_click(&mut self, ui: &mut CalculatorUI) { self.digit(ui, '3'); }
    fn on_n4_click(&mut self, ui: &mut CalculatorUI) { self.digit(ui, '4'); }
    fn on_n5_click(&mut self, ui: &mut CalculatorUI) { self.digit(ui, '5'); }
    fn on_n6_click(&mut self, ui: &mut CalculatorUI) { self.digit(ui, '6'); }
    fn on_n7_click(&mut self, ui: &mut CalculatorUI) { self.digit(ui, '7'); }
    fn on_n8_click(&mut self, ui: &mut CalculatorUI) { self.digit(ui, '8'); }
    fn on_n9_click(&mut self, ui: &mut CalculatorUI) { self.digit(ui, '9'); }
    fn on_dt_click(&mut self, ui: &mut CalculatorUI) { self.digit(ui, '.'); }
    fn on_ad_click(&mut self, ui: &mut CalculatorUI) { self.operation(ui, '+'); }
    fn on_mn_click(&mut self, ui: &mut CalculatorUI) { self.operation(ui, '-'); }
    fn on_ml_click(&mut self, ui: &mut CalculatorUI) { self.operation(ui, '*'); }
    fn on_dv_click(&mut self, ui: &mut CalculatorUI) { self.operation(ui, '/'); }
    fn on_sb_click(&mut self, ui: &mut CalculatorUI) { self.equals(ui); }
    fn on_ac_click(&mut self, ui: &mut CalculatorUI) { self.reset(ui); }
    fn on_dl_click(&mut self, ui: &mut CalculatorUI) { self.delete(ui); }
    fn on_pr_click(&mut self, ui: &mut CalculatorUI) { self.percent(ui); }

    fn handle_key(&mut self, ui: &mut CalculatorUI, key: &str) -> bool
    {
        if key.len() == 1 && key.as_bytes()[0].is_ascii_digit()
        {
            self.digit(ui, key.chars().next().unwrap_or_default());
            return true;
        }
        match key
        {
            "." => self.digit(ui, '.'),
            "+" => self.operation(ui, '+'),
            "-" => self.operation(ui, '-'),
            "*" => self.operation(ui, '*'),
            "/" => self.operation(ui, '/'),
            "=" => self.equals(ui),
            "%" => self.percent(ui),
            "Backspace" => self.delete(ui),
            _ => return false,
        }
        true
    }
}

fn is_operator(value: char) -> bool
{
    matches!(value, '+' | '-' | '*' | '/')
}

fn format_calculator(value: f64) -> String
{
    if value.fract() == 0.0
    {
        return format!("{value:.0}");
    }
    let text = format!("{value:.10}");
    text.trim_end_matches('0').trim_end_matches('.').to_string()
}

fn main()
{
    let code = uimd::run_generated_app_main(||
    {
        let mut ui = CalculatorUI::new();
        let mut app = CalculatorApp::new(&mut ui);
        ui.run(&mut app, &std::env::args().collect::<Vec<_>>())
    });
    std::process::exit(code);
}

#[cfg(test)]
mod tests
{
    use super::*;

    #[test]
    fn calculator_domain_logic_matches_reference()
    {
        let mut ui = CalculatorUI::new();
        let mut app = CalculatorApp::new(&mut ui);
        for key in ["1", "2", "+", "3", "="]
        {
            assert!(app.handle_key(&mut ui, key));
        }
        assert_eq!(ui.display.borrow().text(), "15");
        app.on_ml_click(&mut ui);
        app.on_n2_click(&mut ui);
        app.on_sb_click(&mut ui);
        assert_eq!(ui.display.borrow().text(), "30");
        app.on_pr_click(&mut ui);
        assert_eq!(ui.display.borrow().text(), "0.3");
    }
}
